### DISCUSSION TITLE
search a string in a long text

### DISCUSSION THREAD

**Discussion by uriva** [krisk/Fuse#710] [Q&A] [answered]
hi, I'm trying to use fuse.js to fuzzy search within a large string.

```
const text ='...very large text';

const query = 'some query';

const result = new Fuse([text], {
  threshold: 0.1,
  includeMatches: true,
  includeScore: true,
  isCaseSensitive: false,
  ignoreLocation: true,
}).search(query);

console.log(
  result[0].matches[0].indices.map(([start, end]) => start, end, text.slice(start, end + 1))
);

```

this does match but I don't know which match is best. most are are useless. it's not the longest one, and it's apparently not even the longest one that is shorter than the query.

I think this might require some changes in the source code, to get scoring per match or get them sorted.

Is there any way to achieve this?

Thanks

**Comment by offsky** [comment]
I realize this question is nearly 2 years old. I am also trying to use this library to search large blocks of text and am having the same trouble with the returned matches mostly being useless for highlighting purposes. This screenshot illustrates it.
<img width="605" alt="Screenshot 2024-10-16 at 3 38 47 PM" src="https://github.com/user-attachments/assets/cf96975e-fa6e-43b8-a6de-0c728a428be9">
I have searched for "roots in various cultures" in the source text.  Fuse found the displayed document and gave it a 37% match even though the exact string is found in the text. Fuse also returned 160 matches which I have highlighted. The exact match is highlighted as expected but so are a ton of other useless substrings.  

I haven't dug into the code to understand this yet.  If we could get a score for each individual match, then we could selectively highlight only the good ones.  Or maybe this library isn't built for this use case. I dont know.

**Reply by uriva** [reply]
fyi I ended up using `npm:levenshtein-search` and other complementary approaches (tf-idf)

**Comment by krisk** [comment, marked as answer]
Apologies for the late response.

You're right that this is a limitation — when searching within a large body of text, the `matches` array returns all partial sub-matches without individual scores, making it hard to identify the "best" one for highlighting purposes. The `minMatchCharLength` option can help filter out very short/noisy matches, but it doesn't fully solve the ranking problem.

This is something I'd like to improve. Scoring individual match ranges would make highlighting in long-form text much more useful.

**Reply by uriva** [reply]
ended up building a more complete solution for this since my last comment. it's a strategy cascade: tries levenshtein fuzzy search first (with incrementally increasing threshold so near-exact matches win), then a levenshtein+TF-IDF hybrid, then pure TF-IDF, then exact substring match. for long queries it splits at sentence boundaries, searches each half independently, and combines by proximity.

I use it to find quoted dialog in full movie subtitle files, which are pretty large text blocks.

the fuzzy search part is already on JSR as [@uri/silly-nlp](https://jsr.io/@uri/silly-nlp). if there's interest I can extract the full phrase-in-text search into its own package.
